src/gui/log/logmodel.cpp

```diff
@@ -0,0 +1,173 @@
+{
+    return 1;
+}
+
+QVariant BaseLogModel::data(const QModelIndex &index, const int role) const
+{
+    static int hit = 0, miss = 0;
+    if (!index.isValid() || m_msgCount == 0)
+        return {};
+
+    const int i = m_msgCount - index.row() + m_startIndex - 1;
+
+    Item *item = m_cache[i];
+    if (!item) {
+        item = new Item{rowData(i, Qt::DisplayRole), rowData(i, Qt::ForegroundRole), rowData(i, Qt::UserRole)};
+        m_cache.insert(i, item, 1);
+    }
+
+    switch (role) {
+    case Qt::DisplayRole:
+        return item->displayRole;
+    case Qt::ForegroundRole:
+        return item->foregroundRole;
+    case Qt::UserRole:
+        return item->userRole;
+    default:
+        return {};
+    }
+}
+
+void BaseLogModel::addNewMessage() 
+{
+    beginInsertRows(QModelIndex(), 0, 0);
+    ++m_msgCount;
+    const int i = m_msgCount + m_startIndex - 1;
+    Item *item = new Item{rowData(i, Qt::DisplayRole), rowData(i, Qt::ForegroundRole), rowData(i, Qt::UserRole)};
+    m_cache.insert(i, item, 1);
+    endInsertRows();
+
+    const int count = rowCount();
+    if (count > MAX_LOG_MESSAGES) {
+        const int lastMessage = count - 1;
+        beginRemoveRows(QModelIndex(), lastMessage, lastMessage);
+        --m_msgCount;
+        endRemoveRows();
+    }
+}
+
+void BaseLogModel::reset()
+{
+    beginResetModel();
+    m_startIndex = m_msgCount;
+    m_msgCount = 0;
+    endResetModel();
+}
+
+LogModel::LogModel(QObject * parent) : BaseLogModel(Logger::instance()->messageCount(), parent) 
+{
+    connect(Logger::instance(), &Logger::newLogMessage, this, &LogModel::handleNewMessage);
+}
+
+void LogModel::handleNewMessage(const Log::Msg &)
+{
+    addNewMessage();
+}
+
+QVariant LogModel::rowData(int id, int role) const
+{
+    const Log::Msg msg = Logger::instance()->message(id);
+    switch (role) {
+    case Qt::DisplayRole: {
+        const QDateTime time = QDateTime::fromMSecsSinceEpoch(msg.timestamp);
+        return QString("%1 - %2").arg(time.toString(Qt::SystemLocaleShortDate), msg.message);
+    }
+    case Qt::ForegroundRole:
+        switch (msg.type) {
+        // The RGB QColor constructor is used for performance
+        case Log::INFO:
+            return QColor(0, 0, 255); // blue
+        case Log::WARNING:
+            return QColor(255, 165, 0); // orange
+        case Log::CRITICAL:
+            return QColor(255, 0, 0); // red
+        default:
+            return QApplication::palette().color(QPalette::WindowText);
+        }
+    case Qt::UserRole:
+        return msg.type;
+    }
+
+    return {};
+}
+
+
+LogPeerModel::LogPeerModel(QObject * parent) : BaseLogModel(Logger::instance()->peerCount(), parent) 
+{
+    connect(Logger::instance(), &Logger::newLogPeer, this, &LogPeerModel::handleNewMessage);
+}
+
+void LogPeerModel::handleNewMessage(const Log::Peer &)
+{
+    addNewMessage();
+}
+
+QVariant LogPeerModel::rowData(int id, int role) const
+{
+    const Log::Peer peer = Logger::instance()->peer(id);
```

Comment:
is it guaranteed that `id` will always be valid and will not read outside of `m_peers`? This looks very unsafe to me.

Reply:
`BaseLogModel::data` does the checking, I don't think both inherited classes can do boundary checking (they don't know indexes)